cmd/smith/main.go

```diff
@@ -41,63 +28,9 @@ func runWithContext(ctx context.Context) error {
 	a := app.App{
 		Workers: 2,
 	}
-
-	flag.BoolVar(&a.DisablePodPreset, "disable-pod-preset", false, "Disable PodPreset support")
-	scDisable := flag.Bool("disable-service-catalog", false, "Disable Service Catalog support")
-	scUrl := flag.String("service-catalog-url", "", "Service Catalog API server URL")
-	scInsecure := flag.Bool("service-catalog-insecure", false, "Disable TLS validation for Service Catalog")
-	flag.DurationVar(&a.ResyncPeriod, "resync-period", defaultResyncPeriod, "Resync period for informers")
-	flag.StringVar(&a.Namespace, "namespace", meta_v1.NamespaceAll, "Namespace to use. All namespaces are used if empty string or omitted")
-	pprofAddr := flag.String("pprof-address", "", "Address for pprof to listen on")
-	flag.StringVar(&a.PluginsDir, "plugins-dir", "", "Directory to load plugins from. Defaults to working directory")
-	plugins := flag.String("plugins", "", "Comma separated names of plugins to load")
-	flag.Parse()
-
-	config, err := client.ConfigFromEnv()
-	if err != nil {
-		config, err = rest.InClusterConfig()
-		if err != nil {
-			return err
-		}
-	}
-	config.UserAgent = "smith/" + Version + "/" + GitCommit
-	a.RestConfig = config
-	if !*scDisable {
-		scConfig := *config // shallow copy
-		if *scInsecure {
-			scConfig.TLSClientConfig.Insecure = true
-			scConfig.TLSClientConfig.CAFile = ""
-			scConfig.TLSClientConfig.CAData = nil
-		}
-		if *scUrl != "" {
-			scConfig.Host = *scUrl
-		}
-		a.ServiceCatalogConfig = &scConfig
-	}
-
-	if *plugins != "" {
-		a.Plugins = strings.Split(*plugins, ",")
-	}
-
-	if pprofAddr != nil && *pprofAddr != "" {
-		go func() {
-			log.Fatalf("pprof server failed: %v", http.ListenAndServe(*pprofAddr, nil))
-		}()
+	if err := a.ParseFlags(flag.CommandLine, os.Args[1:]); err != nil {
```

Comment:
I'd probably push as something that automatically happens on Run (and give CommandLine/Args). Just in the interest of simplicity of interface... I don't see any advantage to forcing the plugin author to call ParseFlags before calling run.

In a similar vein, why not just have an 'a.Run' command which sets up 'a' with a default workers of 2? And if you want a different number of workers, have it passed as a flag?

Reply:
1. Originally `app` package did not depend on flags or configuration files or anything like that. It is/was agnostic of where the configuration comes from. This helps a lot when you want to reuse the code or set up the whole app in an integration test. Imagine setting up an integration test with `.Run()` that wants you to pass a FlagSet into it. Ugly. I was actually thinking to put flag parsing into a separate package and keeping `app` clean.
2. Workers should be a command line flag, yes.